### ISSUE TITLE
Make the initial setup more user/beginner friendly

### ISSUE THREAD

**Issue by SamTV12345** [kanidm/kanidm#1726]
**Is your feature request related to a problem? Please describe.**

I have the problem that I can't really get kanidm to start working in a "testing" environment. Keycloak has a page that describes how to start a development container with just one command. Would that be possible to implement in kanidm or have the documentation be extended so I can simply copy & paste the commands. I can learn that way the best

**Describe the solution you'd like**

- Make a one command startup of kanidm with some default settings for a "development" environment.

**Describe alternatives you've considered**

Make the Getting started guide more descriptive so that I can simply copy & paste the commands and have it start once I enter the last command

**Additional context**

[Keycloak Production build](https://www.keycloak.org/server/containers)
[Keycloak one command to get started](https://www.keycloak.org/getting-started/getting-started-docker)

**Comment by yaleman** [comment]
You *can* run a dev environment in two commands - listed [here in the development server docs](https://kanidm.github.io/kanidm/master/DEVELOPER_README.html#development-server-for-interactive-testing). The first command generates the certificates and the second one start it. They're separate for reasons where folks are doing dev, we don't want to rebuild the certs every time you're starting the dev server (I did it probably 30 times this morning, for example.

There's full copy-paste documentation starting [here](https://kanidm.github.io/kanidm/master/installing_the_server.html) - was there something in particular along the way that didn't work for you? We try to keep it as up-to-date as possible, but with our small team we sometimes miss things along the way.

**Comment by Firstyear** [comment]
@SamTV12345 Are you looking for a development environment? Or a quickstart? They really achieve different things.

**Comment by SamTV12345** [comment]
Something like an "evaluation" copy would be cool. With keycloak I first checked if it is suitable for my homelab and it is working stable. With kanidm I first have to install cargo, then generate tls certificates, download all the linked libraries and hope it will start. I get the following error when running `./run_insecure_dev_server.sh recover-account admin`

error[E0599]: no associated item named `NO_CHECK_TIME` found for struct `X509VerifyFlags` in the current scope
    --> /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/webauthn-rs-core-0.4.9/src/attestation.rs:1277:49
     |
1277 |             .set_flags(verify::X509VerifyFlags::NO_CHECK_TIME)
     |                                                 ^^^^^^^^^^^^^
     |                                                 |
     |                                                 associated item not found in `X509VerifyFlags`
     |                                                 help: there is an associated constant with a similar name: `USE_CHECK_TIME`
     
     This is the output of the command:
     
     Caused by:
  process didn't exit successfully: `/root/kanidm/target/debug/build/libssh2-sys-c71e6e977c05cc69/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=LIBSSH2_SYS_USE_PKG_CONFIG
  cargo:include=/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/include
  cargo:root=/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out
  cargo:rerun-if-env-changed=DEP_Z_INCLUDE
  cargo:rerun-if-env-changed=DEP_OPENSSL_INCLUDE
  TARGET = Some("x86_64-alpine-linux-musl")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-alpine-linux-musl")
  cargo:rerun-if-env-changed=CC_x86_64-alpine-linux-musl
  CC_x86_64-alpine-linux-musl = None
  cargo:rerun-if-env-changed=CC_x86_64_alpine_linux_musl
  CC_x86_64_alpine_linux_musl = None
  cargo:rerun-if-env-changed=HOST_CC
  HOST_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64-alpine-linux-musl
  CFLAGS_x86_64-alpine-linux-musl = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_alpine_linux_musl
  CFLAGS_x86_64_alpine_linux_musl = None
  cargo:rerun-if-env-changed=HOST_CFLAGS
  HOST_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/include" "-I" "libssh2/src" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build" "-I" "/root/kanidm/target/debug/build/libz-sys-59b4863c1e11960e/out/include" "-I" "/usr/include" "-fvisibility=hidden" "-DHAVE_LONGLONG" "-DHAVE_SNPRINTF" "-DHAVE_UNISTD_H" "-DHAVE_INTTYPES_H" "-DHAVE_STDLIB_H" "-DHAVE_SYS_SELECT_H" "-DHAVE_SYS_SOCKET_H" "-DHAVE_SYS_IOCTL_H" "-DHAVE_SYS_TIME_H" "-DHAVE_SYS_UN_H" "-DHAVE_O_NONBLOCK" "-DLIBSSH2_OPENSSL" "-DHAVE_LIBCRYPT32" "-DHAVE_EVP_AES_128_CTR" "-DHAVE_POLL" "-DHAVE_GETTIMEOFDAY" "-DLIBSSH2_DH_GEX_NEW" "-DLIBSSH2_HAVE_ZLIB" "-DLIBSSH2DEBUG" "-o" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build/libssh2/src/agent.o" "-c" "libssh2/src/agent.c"
  exit status: 0
  running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/include" "-I" "libssh2/src" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build" "-I" "/root/kanidm/target/debug/build/libz-sys-59b4863c1e11960e/out/include" "-I" "/usr/include" "-fvisibility=hidden" "-DHAVE_LONGLONG" "-DHAVE_SNPRINTF" "-DHAVE_UNISTD_H" "-DHAVE_INTTYPES_H" "-DHAVE_STDLIB_H" "-DHAVE_SYS_SELECT_H" "-DHAVE_SYS_SOCKET_H" "-DHAVE_SYS_IOCTL_H" "-DHAVE_SYS_TIME_H" "-DHAVE_SYS_UN_H" "-DHAVE_O_NONBLOCK" "-DLIBSSH2_OPENSSL" "-DHAVE_LIBCRYPT32" "-DHAVE_EVP_AES_128_CTR" "-DHAVE_POLL" "-DHAVE_GETTIMEOFDAY" "-DLIBSSH2_DH_GEX_NEW" "-DLIBSSH2_HAVE_ZLIB" "-DLIBSSH2DEBUG" "-o" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build/libssh2/src/bcrypt_pbkdf.o" "-c" "libssh2/src/bcrypt_pbkdf.c"
  cargo:warning=libssh2/src/bcrypt_pbkdf.c: In function 'bcrypt_pbkdf':
  cargo:warning=libssh2/src/bcrypt_pbkdf.c:113:24: error: storage size of 'ctx' isn't known
  cargo:warning=  113 |     libssh2_sha512_ctx ctx;
  cargo:warning=      |                        ^~~
  exit status: 1
  running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/include" "-I" "libssh2/src" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build" "-I" "/root/kanidm/target/debug/build/libz-sys-59b4863c1e11960e/out/include" "-I" "/usr/include" "-fvisibility=hidden" "-DHAVE_LONGLONG" "-DHAVE_SNPRINTF" "-DHAVE_UNISTD_H" "-DHAVE_INTTYPES_H" "-DHAVE_STDLIB_H" "-DHAVE_SYS_SELECT_H" "-DHAVE_SYS_SOCKET_H" "-DHAVE_SYS_IOCTL_H" "-DHAVE_SYS_TIME_H" "-DHAVE_SYS_UN_H" "-DHAVE_O_NONBLOCK" "-DLIBSSH2_OPENSSL" "-DHAVE_LIBCRYPT32" "-DHAVE_EVP_AES_128_CTR" "-DHAVE_POLL" "-DHAVE_GETTIMEOFDAY" "-DLIBSSH2_DH_GEX_NEW" "-DLIBSSH2_HAVE_ZLIB" "-DLIBSSH2DEBUG" "-o" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build/libssh2/src/blowfish.o" "-c" "libssh2/src/blowfish.c"
  exit status: 0

  --- stderr
  fatal: not a git repository (or any of the parent directories): .git


  error occurred: Command "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/include" "-I" "libssh2/src" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build" "-I" "/root/kanidm/target/debug/build/libz-sys-59b4863c1e11960e/out/include" "-I" "/usr/include" "-fvisibility=hidden" "-DHAVE_LONGLONG" "-DHAVE_SNPRINTF" "-DHAVE_UNISTD_H" "-DHAVE_INTTYPES_H" "-DHAVE_STDLIB_H" "-DHAVE_SYS_SELECT_H" "-DHAVE_SYS_SOCKET_H" "-DHAVE_SYS_IOCTL_H" "-DHAVE_SYS_TIME_H" "-DHAVE_SYS_UN_H" "-DHAVE_O_NONBLOCK" "-DLIBSSH2_OPENSSL" "-DHAVE_LIBCRYPT32" "-DHAVE_EVP_AES_128_CTR" "-DHAVE_POLL" "-DHAVE_GETTIMEOFDAY" "-DLIBSSH2_DH_GEX_NEW" "-DLIBSSH2_HAVE_ZLIB" "-DLIBSSH2DEBUG" "-o" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build/libssh2/src/bcrypt_pbkdf.o" "-c" "libssh2/src/bcrypt_pbkdf.c" with args "cc" did not execute successfully (status code exit status: 1).

**Comment by SamTV12345** [comment]
> You _can_ run a dev environment in two commands - listed [here in the development server docs](https://kanidm.github.io/kanidm/master/DEVELOPER_README.html#development-server-for-interactive-testing). The first command generates the certificates and the second one start it. They're separate for reasons where folks are doing dev, we don't want to rebuild the certs every time you're starting the dev server (I did it probably 30 times this morning, for example.
> 
> There's full copy-paste documentation starting [here](https://kanidm.github.io/kanidm/master/installing_the_server.html) - was there something in particular along the way that didn't work for you? We try to keep it as up-to-date as possible, but with our small team we sometimes miss things along the way.

There is this step with `docker volume create kanidmd` and in the next step it is stated `You should have a chain.pem and key.pem in your kanidmd volume.` Could you extend that a little more with the concrete commands to execute. Since I'm using a reverse proxy I haven't touched certificates for a long time so a short introduction would be great. Are the certificates used for generating the token? 
If yes would it be possible to make this step an optional task. If I remember correctly Keycloak simply creates these certificates at runtime. This has the drawback that every token is invalidated at a restart but I wouldn't really care about it as it is just me using SSO.

**Comment by Firstyear** [comment]
The certificates are *just* for TLS. All of the internal token signing keys are self-generated at first run. 

Having a section on making certs is a lot of work sadly, there are lots of ways to do it - most of them bad. Generally we'd point to lets encrypt I think. 

I think we can do a better "quick" start though. But at some point it'll have to say "you need tls certs - you need to know how to get these". We are considering adding a CA in the future but that's long term.

**Comment by Firstyear** [comment]
> error occurred: Command "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/include" "-I" "libssh2/src" "-I" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build" "-I" "/root/kanidm/target/debug/build/libz-sys-59b4863c1e11960e/out/include" "-I" "/usr/include" "-fvisibility=hidden" "-DHAVE_LONGLONG" "-DHAVE_SNPRINTF" "-DHAVE_UNISTD_H" "-DHAVE_INTTYPES_H" "-DHAVE_STDLIB_H" "-DHAVE_SYS_SELECT_H" "-DHAVE_SYS_SOCKET_H" "-DHAVE_SYS_IOCTL_H" "-DHAVE_SYS_TIME_H" "-DHAVE_SYS_UN_H" "-DHAVE_O_NONBLOCK" "-DLIBSSH2_OPENSSL" "-DHAVE_LIBCRYPT32" "-DHAVE_EVP_AES_128_CTR" "-DHAVE_POLL" "-DHAVE_GETTIMEOFDAY" "-DLIBSSH2_DH_GEX_NEW" "-DLIBSSH2_HAVE_ZLIB" "-DLIBSSH2DEBUG" "-o" "/root/kanidm/target/debug/build/libssh2-sys-6f332baad66e0713/out/build/libssh2/src/bcrypt_pbkdf.o" "-c" "libssh2/src/bcrypt_pbkdf.c" with args "cc" did not execute successfully (status code exit status: 1).

There is some proper chaos going on here. But yeah, I don't think the insecure dev server is what you want for "evaluation".

**Comment by yaleman** [comment]
> rust errors

I'd probably say try the docker container if you want a quick-start in that case - it looks like you're missing the [dependencies listed in the Developer readme](https://kanidm.github.io/kanidm/stable/DEVELOPER_README.html).

> There is this step with `docker volume create kanidmd` and in the next step it is stated `You should have a chain.pem and key.pem in your kanidmd volume.` Could you extend that a little more with the concrete commands to execute. Since I'm using a reverse proxy I haven't touched certificates for a long time so a short introduction would be great. Are the certificates used for generating the token? If yes would it be possible to make this step an optional task. If I remember correctly Keycloak simply creates these certificates at runtime. This has the drawback that every token is invalidated at a restart but I wouldn't really care about it as it is just me using SSO.

Copying the certificates into the volume is a good catch, thanks. The certificates are used for all communications with Kanidm, see ["Why TLS" in our FAQ](https://kanidm.github.io/kanidm/stable/frequently_asked_questions.html?highlight=tls#why-tls). Auth tokens are signed with an internally managed signing key.

We really don't want to be automatically generating the *certificates* as there's plenty of other systems out there to handle cert generation, and reinventing the wheel isn't a good use of dev resources.

So, to summarise we need to add:

- a reference to [letsencrypt](https://letsencrypt.org) and [cfssl](https://github.com/cloudflare/cfssl) for generating certs
- `docker cp` commands for copying the certificate into the volume

**Comment by SamTV12345** [comment]
Yeah that would be great. 👍🏻 I Never really used docker cp before.
